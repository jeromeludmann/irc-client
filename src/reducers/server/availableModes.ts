import { Reducer } from "redux";
import { ReplyMyInfoReceivedAction, RPL_MYINFO_RECEIVED } from "@app/actions/messages";

export type AvailableServerModesState = {
  user: string[];
  channel: string[];
};

export type AvailableModesAction = ReplyMyInfoReceivedAction;

export const availableServerModesInitialState: AvailableServerModesState = {
  user: [],
  channel: []
};

const handlers: { [action: string]: Reducer<AvailableServerModesState> } = {
  [RPL_MYINFO_RECEIVED]: (_, action: ReplyMyInfoReceivedAction) => ({
    user: action.payload.availableUserModes,
    channel: action.payload.availableChannelModes
  })
};

export const reduceAvailableServerModes = (
  availableModes = availableServerModesInitialState,
  action: AvailableModesAction
) =>
  handlers.hasOwnProperty(action.type)
    ? handlers[action.type](availableModes, action)
    : availableModes;
