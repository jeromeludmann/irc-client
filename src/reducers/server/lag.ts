import {
  PongFromServerReceivedAction,
  PONG_FROM_SERVER_RECEIVED,
} from "@app/actions/messages";
import { Reducer, Action } from "redux";

export type ServerLagState = number;

export const serverLagInitialState = 0;

const handlers: { [action: string]: Reducer<ServerLagState> } = {
  [PONG_FROM_SERVER_RECEIVED]: (
    _serverLag,
    action: PongFromServerReceivedAction,
  ) => action.payload.lag,
};

export const reduceLag = (
  serverLag: ServerLagState = serverLagInitialState,
  action: Action,
): ServerLagState =>
  handlers.hasOwnProperty(action.type)
    ? handlers[action.type](serverLag, action)
    : serverLag;
