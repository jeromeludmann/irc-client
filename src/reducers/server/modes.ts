export type ServerModesState = string[];

export const serverModesInitialState: ServerModesState = [];

export const reduceServerModes = (
  modes = serverModesInitialState,
): ServerModesState => modes;
