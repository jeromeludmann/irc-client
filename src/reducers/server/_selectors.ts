import { createSelector } from "reselect";
import { selectServer, selectRoute } from "@app/reducers/_selectors";

export const selectUser = createSelector(selectServer, server => server.user);

export const selectChannel = createSelector(
  selectServer,
  selectRoute,
  (server, active) => server.channels[active.channelKey],
);
