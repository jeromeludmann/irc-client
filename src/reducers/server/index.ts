import {
  ServerModesState,
  serverModesInitialState,
  reduceServerModes,
} from "@app/reducers/server/modes";
import { RouteState } from "@app/reducers/route";
import {
  ChannelsState,
  channelsInitialState,
  reduceChannels,
} from "@app/reducers/server/channels";
import { RoutedAction } from "@app/Route";
import {
  serverNameInitialState,
  reduceServerName,
} from "@app/reducers/server/name";
import {
  reduceAvailableServerModes,
  AvailableServerModesState,
  availableServerModesInitialState,
  AvailableModesAction,
} from "@app/reducers/server/availableModes";
import { ServerNameState } from "@app/reducers/server/name";
import {
  UserState,
  userInitialState,
  reduceUser,
} from "@app/reducers/server/user";
import { PongFromServerReceivedAction } from "@app/actions/messages";
import {
  reduceLag,
  ServerLagState,
  serverLagInitialState,
} from "@app/reducers/server/lag";

export type ServerState = {
  name: ServerNameState;
  user: UserState;
  availableModes: AvailableServerModesState;
  modes: ServerModesState;
  channels: ChannelsState;
  lag: ServerLagState;
};

export const serverInitialState: ServerState = {
  name: serverNameInitialState,
  user: userInitialState,
  availableModes: availableServerModesInitialState,
  modes: serverModesInitialState,
  channels: channelsInitialState,
  lag: serverLagInitialState,
};

export const reduceServer = (
  server = serverInitialState,
  action: RoutedAction,
  extraStates: { route: RouteState },
): ServerState => ({
  name: reduceServerName(server.name, action),
  user: reduceUser(server.user, action),
  availableModes: reduceAvailableServerModes(
    server.availableModes,
    action as AvailableModesAction,
  ),
  modes: reduceServerModes(server.modes),
  channels: reduceChannels(server.channels, action, {
    ...extraStates,
    user: server.user,
  }),
  lag: reduceLag(server.lag, action as PongFromServerReceivedAction),
});
