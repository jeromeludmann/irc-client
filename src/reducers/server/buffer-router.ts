import { Action } from "redux";
import reduceBuffer, {
  BufferState,
  bufferInitialState,
} from "@app/reducers/buffer/buffer";
import {
  Route,
  STATUS_BUFFER,
  RAW_BUFFER,
  ALL_BUFFERS,
  NO_BUFFER,
} from "@app/Route";
import { WindowState } from "@app/reducers/window";
import { UserState } from "@app/reducers/user";

export interface BufferRouterState {
  readonly [key: string]: BufferState;
}

export type BufferRouterAction = Action;

interface ExtraParams {
  readonly user: UserState;
  readonly route: Route;
  readonly active: WindowState;
}

export const bufferRouterInitialState: BufferRouterState = {
  [RAW_BUFFER]: bufferInitialState,
  [STATUS_BUFFER]: bufferInitialState,
};

const registry: {
  [bufferKey: string]: (
    bufferRouter: BufferRouterState,
    action: BufferRouterAction,
    params: ExtraParams,
  ) => BufferRouterState;
} = {
  [NO_BUFFER](bufferRouter) {
    return bufferRouter;
  },
  [ALL_BUFFERS](bufferRouter, action, params) {
    const buffers: { [key: string]: BufferState } = {};
    Object.keys(bufferRouter).forEach(key => {
      if (key !== RAW_BUFFER) {
        buffers[key] = reduceBuffer(bufferRouter[key], action, params);
      }
    });
    return { ...bufferRouter, ...buffers };
  },
};

export default (
  bufferRouter = bufferRouterInitialState,
  action: BufferRouterAction,
  params: ExtraParams,
): BufferRouterState => {
  const bufferKey = params.route.bufferKey;

  if (registry.hasOwnProperty(bufferKey)) {
    return registry[bufferKey](bufferRouter, action, params);
  }

  return bufferKey
    ? {
        ...bufferRouter,
        [bufferKey]: reduceBuffer(bufferRouter[bufferKey], action, params),
      }
    : bufferRouter;
};
