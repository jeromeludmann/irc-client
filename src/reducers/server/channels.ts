import {
  ChannelState,
  channelInitialState,
  reduceChannel,
} from "@app/reducers/channel";
import {
  STATUS,
  RAW,
  BROADCAST_ALL,
  BROADCAST_NONE,
  RoutedAction,
  isRaw,
  BROADCAST_ACTIVE,
  isPrivate,
  isChannel,
} from "@app/Route";
import { RouteState } from "@app/reducers/route";
import { UserState } from "@app/reducers/server/user";
import { CLOSE_WINDOW, CloseWindowAction } from "@app/actions/ui";
import { JOIN, PRIVMSG } from "@app/actions/messages";

export interface ChannelsState {
  readonly [key: string]: ChannelState;
}

export const channelsInitialState: ChannelsState = {
  [RAW]: channelInitialState,
  [STATUS]: channelInitialState,
};

type ChannelsReducer<A = RoutedAction> = (
  channels: ChannelsState,
  action: A,
  extraStates: { route: RouteState; user: UserState },
) => ChannelsState;

const closeWindow: ChannelsReducer<CloseWindowAction> = (channels, action) => {
  const updatedChannels = { ...channels };

  if (
    isChannel(action.route.channelKey) ||
    isPrivate(action.route.channelKey)
  ) {
    delete updatedChannels[action.route.channelKey];
    return updatedChannels;
  }

  Object.keys(channels).forEach(channelKey => {
    if (isChannel(channelKey) || isPrivate(channelKey)) {
      delete updatedChannels[channelKey];
    }
  });

  return updatedChannels;
};

// const join: ChannelsReducer<JoinAction> = (channels, action, extraStates) => ({
//   ...channels,
//   [action.payload.channel]: reduceChannel(
//     channels[action.payload.channel],
//     action,
//     extraStates,
//   ),
// });

// const privmsg: ChannelsReducer<PrivmsgAction> = (
//   channels,
//   action,
//   extraStates,
// ) => {
//   if (!isChannel(action.route.channelKey)) {
//     // create channel key node
//   }
//   if (!channels.hasOwnProperty(action.route.channelKey)) {
//     return {
//       ...channels,
//       [action.route.channelKey]: reduceChannel(
//         channels[action.route.channelKey],
//         action,
//         extraStates,
//       ),
//     };
//   }

//   return channels;
// };

const broadcastActive: ChannelsReducer = (channels, action, extraStates) => {
  const key = extraStates.route.channelKey;
  return {
    ...channels,
    [key]: reduceChannel(channels[key], action, extraStates),
  };
};

const broadcastAll: ChannelsReducer = (channels, action, extraStates) => {
  const broadcastedChannels: { [key: string]: ChannelState } = {};
  Object.keys(channels).forEach(key => {
    if (!isRaw(key)) {
      broadcastedChannels[key] = reduceChannel(
        channels[key],
        action,
        extraStates,
      );
    }
  });
  return { ...channels, ...broadcastedChannels };
};

const map: { [action: string]: ChannelsReducer } = {
  [CLOSE_WINDOW]: closeWindow,
};

const routes: { [channelKey: string]: ChannelsReducer } = {
  [BROADCAST_NONE]: channels => channels,
  [BROADCAST_ACTIVE]: broadcastActive,
  [BROADCAST_ALL]: broadcastAll,
};

export const reduceChannels = (
  channels = channelsInitialState,
  action: RoutedAction,
  extraStates: { route: RouteState; user: UserState },
): ChannelsState => {
  const channelKey = action.route.channelKey;

  if (!channelKey) {
    return channels;
  }

  if (routes.hasOwnProperty(channelKey)) {
    return routes[channelKey](channels, action, extraStates);
  }

  if (map.hasOwnProperty(action.type)) {
    return map[action.type](channels, action, extraStates);
  }

  if (
    channels.hasOwnProperty(channelKey) ||
    action.type === PRIVMSG ||
    action.type === JOIN
  ) {
    return {
      ...channels,
      [channelKey]: reduceChannel(channels[channelKey], action, extraStates),
    };
  }

  return channels;
};
