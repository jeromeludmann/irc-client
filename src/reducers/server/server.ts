import reduceModes, {
  ModesState,
  modesInitialState,
} from "@app/reducers/server/modes";
import { WindowState } from "@app/reducers/window";
import reduceBufferRouter, {
  BufferRouterState,
  bufferRouterInitialState,
  BufferRouterAction,
} from "@app/reducers/server/buffer-router";
import { Route } from "@app/Route";
import { UserState } from "@app/reducers/user";
import { RPL_MYINFO, IncomingReplyMyInfoAction } from "@app/actions/incoming";
import { Action } from "redux";

export type ServerState = {
  readonly name: string; // TODO
  readonly availableModes: { user: string[]; channel: string[] }; // TODO
  readonly modes: ModesState;
  readonly buffers: BufferRouterState;
};

export type ServerAction = Action;

interface ExtraParams {
  user: UserState;
  route: Route;
  active: WindowState;
}

export const serverInitialState: ServerState = {
  name: "<unknown>", // TODO
  availableModes: { user: [], channel: [] }, // TODO
  modes: modesInitialState,
  buffers: bufferRouterInitialState,
};

// TODO extract me
function reduceName(name: string, action: IncomingReplyMyInfoAction) {
  switch (action.type) {
    case RPL_MYINFO:
      return action.payload.serverName;
    default:
      return name;
  }
}

// TODO extract me
function reduceAvailableModes(
  availableModes: any,
  action: IncomingReplyMyInfoAction,
) {
  switch (action.type) {
    case RPL_MYINFO:
      return {
        user: action.payload.availableUserModes,
        channel: action.payload.availableChannelModes,
      };
    default:
      return availableModes;
  }
}

export default (
  server = serverInitialState,
  action: ServerAction,
  params: ExtraParams,
): ServerState => ({
  name: reduceName(server.name, action as IncomingReplyMyInfoAction),
  availableModes: reduceAvailableModes(
    server.availableModes,
    action as IncomingReplyMyInfoAction,
  ),
  modes: reduceModes(server.modes),
  buffers: reduceBufferRouter(
    server.buffers,
    action as BufferRouterAction,
    params,
  ),
});
