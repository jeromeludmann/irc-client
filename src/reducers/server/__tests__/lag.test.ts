import { reduceLag } from "@app/reducers/server/lag";
import { messagesReceived } from "@app/actions/messages";

test("reduce server lag", () => {
  expect(
    reduceLag(
      undefined,
      messagesReceived["PONG"]("server1", "irc.network", ["server1", "key"]),
    ),
  ).toMatchSnapshot();
});
