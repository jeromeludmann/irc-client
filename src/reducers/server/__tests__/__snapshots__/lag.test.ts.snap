// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`reduce server lag 1`] = `0`;
